### DISCUSSION TITLE
Microkernels and RTOSs

### DISCUSSION THREAD

**Discussion by edwardalee** [lf-lang/lingua-franca#1434] [Ideas]
A few of us have been discussing and experimenting with LF on bare iron, on microkernels, and on RTOSs. This discussion can become a place to collect experiences and ideas.

## Bare Iron

Running LF on bare iron really means using a board vendor's libraries for accessing devices such as timers, serial ports, I/O busses, and networking interfaces. So far, we have built the following demonstrators:

* Arduino: Supported directly by the C target in LF. **FIXME**: More info here.
* nrF52x: Supporting boards from Nordic Semiconductor. The realization is in the [lf-buckler](https://github.com/icyphy/lf-buckler) repo. It uses a [particular version of an SDK from Nordic Semiconductor](https://infocenter.nordicsemi.com/index.jsp?topic=%2Fcom.nordic.infocenter.sdk5.v15.3.0%2Findex.html). A suite of demonstrator projects are described in the [EECS 149 lab manual](https://github.com/icyphy/eecs149-lab-book).

A significant downside of the bare-iron approach is that each board may require a significant amount of effort to support. Also, building a threaded implementation may be particularly challenging.

## Microkernels

Microkernels may provide more portable support for threaded execution as well as network interfaces and timing services. Of particular interest:

* [sel4](https://sel4.systems): Open source (GPL) microkernel supporting several architectures (x86, ARM, RISC-V) with drivers for a variety of specific realizations. Includes support for threads, timers, network stacks, and file systems, mostly as add-ons to a basic kernel.

## RTOSs

RTOSs promise more portable code in the sense that the same LF runtime may execute on a variety of platforms. However, LF timing semantics may not be a good match for RTOS timing semantics. Of particular interest:

* Zephyr
* FreeRTOS

**Comment by schoeberl** [comment]
I would like to add that we had LF running on bare iron with Patmos. We also mapped multithreading to multiple cores. The maximum number of threads was the maximum number of cores. We did not add any single-core thread switching code. for the timer interface and the threads, we used an implementation of the phtread standard.
